fontscan/fontmap.go

```diff
@@ -0,0 +1,489 @@
+package fontscan
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+
+	"github.com/go-text/typesetting/font"
+	meta "github.com/go-text/typesetting/opentype/api/metadata"
+	"github.com/go-text/typesetting/opentype/loader"
+)
+
+type cacheEntry struct {
+	Location
+
+	Family string
+	meta.Aspect
+}
+
+// The family substitution algorithm is copied from fontconfig
+// and the match algorithm is inspired from Rust font-kit library
+
+// FontMap provides a mechanism to select a [font.Face] from a font description.
+// It supports system and user-provided fonts, and implements the CSS font substitutions
+// rules.
+//
+// A typical usage would be as following :
+//
+//	fontMap := NewFontMap("cachedir")
+//
+//	// at least one of the following calls
+//	fontMap.UseSystemFonts() // error handling omitted
+//	fontMap.AddFont(font1, "font1") // error handling omitted
+//	fontMap.AddFace(alreadyParsed, metadata) // error handling omitted
+//
+//	// set the font description
+//	fontMap.SetQuery(Query{Families: []string{"Arial", "serif"}}) // regular Aspect
+//
+//	// `fontMap` is now ready for text shaping, using the `ResolveFace` method
+//
```

Comment:
We should convert this into a proper example. The API has already drifted so that this isn't valid at least once.